test/staking/AbstractStaking.test.ts

```diff
@@ -0,0 +1,434 @@
+      expect(await abstractStaking.userOwedValue(FIRST)).to.equal(wei(3));
+
+      await abstractStaking.connect(FIRST).claim((await abstractStaking.getOwedValue(FIRST)) - wei(2));
+
+      expect(await abstractStaking.getOwedValue(FIRST)).to.equal(wei(2));
+      expect(await abstractStaking.userOwedValue(FIRST)).to.equal(wei(2));
+
+      await abstractStaking.connect(FIRST).claim(wei(2));
+
+      expect(await abstractStaking.getOwedValue(FIRST)).to.equal(0);
+      expect(await abstractStaking.userOwedValue(FIRST)).to.equal(wei(0));
+    });
+
+    it("should transfer tokens correctly on the claim", async () => {
+      await performStakingManipulations();
+
+      const initialRewardsBalance = await REWARDS_TOKEN.balanceOf(abstractStaking);
+
+      const firstOwed = await abstractStaking.getOwedValue(FIRST);
+      const secondOwed = await abstractStaking.getOwedValue(SECOND);
+      const thirdOwed = await abstractStaking.getOwedValue(THIRD);
+
+      await abstractStaking.connect(FIRST).claim(firstOwed);
+      await abstractStaking.connect(SECOND).claim(secondOwed);
+
+      await expect(abstractStaking.connect(THIRD).claim(thirdOwed))
+        .to.emit(REWARDS_TOKEN, "Transfer")
+        .withArgs(await abstractStaking.getAddress(), THIRD.address, thirdOwed);
+
+      expect(await REWARDS_TOKEN.balanceOf(abstractStaking)).to.equal(
+        initialRewardsBalance - (firstOwed + secondOwed + thirdOwed),
+      );
+      expect(await REWARDS_TOKEN.balanceOf(FIRST)).to.equal(firstOwed);
+      expect(await REWARDS_TOKEN.balanceOf(SECOND)).to.equal(secondOwed);
+      expect(await REWARDS_TOKEN.balanceOf(THIRD)).to.equal(thirdOwed);
+    });
+
+    it("should not allow to claim 0 rewards", async () => {
+      await expect(abstractStaking.connect(FIRST).claim(0)).to.be.revertedWith(
+        "ValueDistributor: amount has to be more than 0",
+      );
+    });
+
+    it("should not allow to claim more rewards than earned", async () => {
+      await performStakingManipulations();
+
+      await expect(abstractStaking.connect(FIRST).claim(wei(4))).to.be.revertedWith(
+        "ValueDistributor: insufficient amount",
+      );
+    });
+  });
+
+  describe("should handle staking manipulations with 6-decimal values", () => {
+    it("should handle the whole staling process using 6-decimal values", async () => {
+      const AbstractStakingMock = await ethers.getContractFactory("AbstractStakingMock");
+      const ERC20Mock = await ethers.getContractFactory("ERC20Mock");
+
+      const abstractStaking6Decimals = await AbstractStakingMock.deploy();
+      const SHARES_TOKEN = await ERC20Mock.deploy("SharesMock", "SMock", 6);
+      const REWARDS_TOKEN = await ERC20Mock.deploy("RewardsMock", "RMock", 6);
+      await REWARDS_TOKEN.mint(await abstractStaking6Decimals.getAddress(), wei(1000));
+
+      await abstractStaking6Decimals.__AbstractStakingMock_init(SHARES_TOKEN, REWARDS_TOKEN, wei(1, 6), 3n);
```

Comment:
You can use an existing abstract staking contract for that. Then call performStakingManipulations